File: server/app/api/v1/endpoints/creations.py
"""
Creations API Endpoint
Handles creation of user greeting cards and retrieval.
Aligned with the new `creations` table schema (005_destructive_reset).

Security: Authenticated endpoints use ``get_rls_client`` so every Supabase
query runs under the user's JWT (Postgres RLS enforced).
"""

from datetime import datetime, timedelta, timezone
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from supabase import Client

from app.core import CurrentUser, get_settings, verify_user, get_rls_client
from app.core.color_palette import ALLOWED_COLORS

router = APIRouter()

# Maximum length for a URL stored in metadata (guard against base64)
MAX_URL_LENGTH = 2048


# =============================================================================
# VALIDATION HELPERS
# =============================================================================

def validate_metadata(
    metadata: dict[str, Any],
    config_schema: dict[str, Any],
) -> list[str]:
    """
    Validate metadata against the template's config_schema.
    Returns a list of validation error strings (empty = valid).
    """
    errors: list[str] = []
    fields = config_schema.get("fields", [])

    for field_def in fields:
        if not isinstance(field_def, dict):
            continue

        key = field_def.get("key", "")
        field_type = field_def.get("type", "")
        value = metadata.get(key)

        # Required check
        if field_def.get("required") and (value is None or value == "" or value == []):
            errors.append(f"Field '{key}' is required.")
            continue

        if value is None:
            continue

        # Color validation
        if field_type == "color":
            if not isinstance(value, str) or value.upper() not in ALLOWED_COLORS:
                errors.append(
                    f"Field '{key}' value '{value}' is not in the allowed color palette."
                )

        # Image URL validation — reject base64 blobs
        if field_type == "image_url":
            if not isinstance(value, str):
                errors.append(f"Field '{key}' must be a string URL.")
            elif len(value) > MAX_URL_LENGTH:
                errors.append(
                    f"Field '{key}' value is too long ({len(value)} chars). "
                    f"Max {MAX_URL_LENGTH}. Do not embed base64 data."
                )
            elif value.startswith("data:"):
                errors.append(
                    f"Field '{key}' must be a URL, not an inline data URI."
                )

    return errors


# =============================================================================
# REQUEST / RESPONSE MODELS
# =============================================================================

class CreateCreationRequest(BaseModel):
    """Request to create a new creation (greeting card)."""
    template_id: str
    metadata: dict[str, Any]


class CreateCreationResponse(BaseModel):
    """Response after creating a creation."""
    id: str
    expires_at: str | None


class CreationListItem(BaseModel):
    """Creation item for listing user's creations."""
    id: str
    template_slug: str
    template_name: str
    is_paid: bool | None
    expires_at: str | None
    is_deleted: bool
    created_at: str


class CreationDetailResponse(BaseModel):
    """Public creation detail response."""
    id: str
    template_slug: str
    template_name: str
    metadata: dict[str, Any]
    is_paid: bool | None
    expires_at: str | None
    created_at: str


# =============================================================================
# ENDPOINTS
# =============================================================================

@router.post("", response_model=CreateCreationResponse)
async def create_creation(
    request: CreateCreationRequest,
    current_user: CurrentUser = Depends(verify_user),
    supabase: Client = Depends(get_rls_client),
):
    """
    Create a new creation (greeting card).

    - Fetches template to get config_schema & expiration_policy
    - Validates metadata against config_schema
    - Checks creations_left quota based on user tier
    - Decrements creations_left inside a transaction-like sequence
    - Calculates expires_at from template expiration_policy
    - Inserts into creations table
    """
    settings = get_settings()

    # 1. Get template info
    template_result = (
        supabase.table("templates")
        .select("id, slug, name, is_premium, config_schema, expiration_policy")
        .eq("id", request.template_id)
        .eq("is_active", True)
        .single()
        .execute()
    )

    if not template_result.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Template not found",
        )

    template = template_result.data

    # 2. Validate metadata against config_schema
    config_schema = template.get("config_schema") or {}
    validation_errors = validate_metadata(request.metadata, config_schema)
    if validation_errors:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Validation errors: {'; '.join(validation_errors)}",
        )

    # 3. Get user profile to check quota
    profile_result = (
        supabase.table("profiles")
        .select("subscription_tier, creations_count, creations_left_free, creations_left_pro")
        .eq("id", current_user.id)
        .single()
        .execute()
    )

    if not profile_result.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Profile not found. Please complete registration.",
        )

    profile = profile_result.data
    user_tier = profile.get("subscription_tier", "free")

    # 4. Premium template check
    if template.get("is_premium", False) and user_tier == "free":
        raise HTTPException(
            status_code=status.HTTP_402_PAYMENT_REQUIRED,
            detail="TEMPLATE_NOT_ALLOWED",
        )

    # 5. Check creations_left quota
    if user_tier == "free":
        creations_left = profile.get("creations_left_free", 0)
        if creations_left <= 0:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="QUOTA_EXCEEDED",
            )
    elif user_tier == "premium":
        creations_left = profile.get("creations_left_pro")
        # NULL means unlimited for premium
        if creations_left is not None and creations_left <= 0:
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="QUOTA_EXCEEDED",
            )

    # 6. Calculate expiry
    is_paid = user_tier == "premium"
    expiration_policy = template.get("expiration_policy") or {}
    if is_paid:
        expiry_days = expiration_policy.get("paid_days", 14)
    else:
        expiry_days = expiration_policy.get("free_days", 1)

    expires_at = datetime.now(timezone.utc) + timedelta(days=int(expiry_days))

    # 7. Insert creation
    creation_result = (
        supabase.table("creations")
        .insert({
            "user_id": current_user.id,
            "template_id": request.template_id,
            "metadata": request.metadata,
            "is_paid": is_paid,
            "expires_at": expires_at.isoformat(),
        })
        .execute()
    )

    if not creation_result.data:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create card",
        )

    creation = creation_result.data[0]

    # 8. Decrement creations_left and increment creations_count
    update_dict: dict[str, Any] = {
        "creations_count": (profile.get("creations_count", 0) + 1),
    }
    if user_tier == "free":
        update_dict["creations_left_free"] = max(profile.get("creations_left_free", 0) - 1, 0)
    elif user_tier == "premium" and profile.get("creations_left_pro") is not None:
        update_dict["creations_left_pro"] = max(profile.get("creations_left_pro", 0) - 1, 0)

    supabase.table("profiles").update(update_dict).eq("id", current_user.id).execute()

    return CreateCreationResponse(
        id=creation["id"],
        expires_at=creation.get("expires_at"),
    )


@router.get("/my", response_model=list[CreationListItem])
async def get_my_creations(
    current_user: CurrentUser = Depends(verify_user),
    supabase: Client = Depends(get_rls_client),
):
    """Get current user's creations (RLS enforced)."""
    result = (
        supabase.table("creations")
        .select("id, is_paid, expires_at, is_deleted, created_at, templates!inner(slug, name)")
        .eq("user_id", current_user.id)
        .eq("is_deleted", False)
        .order("created_at", desc=True)
        .execute()
    )

    items = []
    for c in result.data or []:
        tmpl = c.get("templates") or {}
        items.append(CreationListItem(
            id=c["id"],
            template_slug=tmpl.get("slug", ""),
            template_name=tmpl.get("name", "כרטיס"),
            is_paid=c.get("is_paid"),
            expires_at=c.get("expires_at"),
            is_deleted=c.get("is_deleted", False),
            created_at=c["created_at"],
        ))

    return items


@router.get("/{creation_id}", response_model=CreationDetailResponse)
async def get_creation(creation_id: str):
    """
    Get a creation by ID (PUBLIC endpoint for shared links).
    Uses anon client — RLS allows SELECT on non-deleted creations.
    """
    from supabase import create_client
    settings = get_settings()
    supabase = create_client(settings.supabase_url, settings.supabase_anon_key)

    result = (
        supabase.table("creations")
        .select("id, metadata, is_paid, expires_at, is_deleted, created_at, templates!inner(slug, name)")
        .eq("id", creation_id)
        .single()
        .execute()
    )

    if not result.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Creation not found",
        )

    c = result.data

    if c.get("is_deleted"):
        raise HTTPException(
            status_code=status.HTTP_410_GONE,
            detail="This creation has been deleted",
        )

    # Check expiry
    if c.get("expires_at"):
        exp_str = c["expires_at"]
        try:
            if "Z" in exp_str:
                exp_dt = datetime.fromisoformat(exp_str.replace("Z", "+00:00"))
            else:
                exp_dt = datetime.fromisoformat(exp_str)
            if exp_dt < datetime.now(timezone.utc):
                raise HTTPException(
                    status_code=status.HTTP_410_GONE,
                    detail="This creation has expired",
                )
        except (ValueError, TypeError):
            pass

    tmpl = c.get("templates") or {}

    return CreationDetailResponse(
        id=c["id"],
        template_slug=tmpl.get("slug", ""),
        template_name=tmpl.get("name", "כרטיס"),
        metadata=c.get("metadata") or {},
        is_paid=c.get("is_paid"),
        expires_at=c.get("expires_at"),
        created_at=c["created_at"],
    )


@router.delete("/{creation_id}")
async def delete_creation(
    creation_id: str,
    current_user: CurrentUser = Depends(verify_user),
    supabase: Client = Depends(get_rls_client),
):
    """Soft delete a creation (owner only, RLS enforced)."""
    result = (
        supabase.table("creations")
        .select("id")
        .eq("id", creation_id)
        .eq("user_id", current_user.id)
        .execute()
    )

    if not result.data:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Creation not found",
        )

    supabase.table("creations").update({"is_deleted": True}).eq("id", creation_id).execute()

    return {"message": "Creation deleted successfully"}

File: client/src/actions/creations.ts
/**
 * Creations Server Actions
 *
 * Port of:
 *   server/app/api/v1/endpoints/creations.py
 *
 * DB columns (creations table – 005_destructive_reset.sql):
 *   id, user_id, template_id, metadata, is_paid, expires_at, is_deleted, created_at
 *
 * DB columns (profiles table – used for quota checks):
 *   subscription_tier, creations_count, creations_left_free, creations_left_pro
 *
 * DB columns (templates table – used for config_schema & expiration_policy):
 *   id, slug, name, is_premium, config_schema, is_active, expiration_policy
 */

"use server";

import { createClient } from "@/lib/supabase/server";
import {
  CreateCreationRequestSchema,
  type CreateCreationInput,
  type CreateCreationResponse,
  type CreationListItem,
  type CreationDetail,
} from "@/lib/validations";
import { validateMetadata } from "@/lib/validations/metadata";

// ---------------------------------------------------------------------------
// CREATE
// ---------------------------------------------------------------------------

/**
 * POST /creations → createCreation(input)
 *
 * Business logic ported from Python endpoint step-by-step:
 *  1. Fetch template (config_schema, expiration_policy, is_premium)
 *  2. Validate metadata against config_schema
 *  3. Fetch user profile for quota check
 *  4. Premium template guard
 *  5. creations_left quota check
 *  6. Calculate expires_at from template expiration_policy
 *  7. Insert creation
 *  8. Decrement creations_left / increment creations_count
 */
export async function createCreation(
  input: CreateCreationInput,
): Promise<
  { data: CreateCreationResponse } | { error: string; status: number }
> {
  try {
    // Zod validation (mirrors Pydantic CreateCreationRequest)
    const parsed = CreateCreationRequestSchema.safeParse(input);
    if (!parsed.success) {
      return {
        error: parsed.error.issues.map((i) => i.message).join("; "),
        status: 422,
      };
    }

    const supabase = await createClient();

    const {
      data: { user },
      error: authError,
    } = await supabase.auth.getUser();

    if (authError || !user) {
      return { error: "Unauthorized", status: 401 };
    }

    // ── Step 1: Get template info ──────────────────────────────────
    const { data: template, error: tmplErr } = await supabase
      .from("templates")
      .select("id, slug, name, is_premium, config_schema, expiration_policy")
      .eq("id", parsed.data.template_id)
      .eq("is_active", true)
      .single();

    if (tmplErr || !template) {
      return { error: "Template not found", status: 404 };
    }

    // ── Step 2: Validate metadata against config_schema ────────────
    const configSchema = (template.config_schema as Record<string, unknown>) ?? {};
    const validationErrors = validateMetadata(
      parsed.data.metadata as Record<string, unknown>,
      configSchema,
    );
    if (validationErrors.length > 0) {
      return {
        error: `Validation errors: ${validationErrors.join("; ")}`,
        status: 422,
      };
    }

    // ── Step 3: Get user profile to check quota ────────────────────
    const { data: profile, error: profErr } = await supabase
      .from("profiles")
      .select(
        "subscription_tier, creations_count, creations_left_free, creations_left_pro",
      )
      .eq("id", user.id)
      .single();

    if (profErr || !profile) {
      return {
        error: "Profile not found. Please complete registration.",
        status: 404,
      };
    }

    const userTier: string = profile.subscription_tier ?? "free";

    // ── Step 4: Premium template check ─────────────────────────────
    if (template.is_premium && userTier === "free") {
      return { error: "TEMPLATE_NOT_ALLOWED", status: 402 };
    }

    // ── Step 5: Check creations_left quota ─────────────────────────
    if (userTier === "free") {
      const creationsLeft: number = profile.creations_left_free ?? 0;
      if (creationsLeft <= 0) {
        return { error: "QUOTA_EXCEEDED", status: 403 };
      }
    } else if (userTier === "premium") {
      const creationsLeft: number | null = profile.creations_left_pro;
      // NULL means unlimited for premium
      if (creationsLeft !== null && creationsLeft <= 0) {
        return { error: "QUOTA_EXCEEDED", status: 403 };
      }
    }

    // ── Step 6: Calculate expiry ───────────────────────────────────
    const isPaid = userTier === "premium";
    const expirationPolicy =
      (template.expiration_policy as Record<string, unknown>) ?? {};

    const expiryDays = isPaid
      ? Number(expirationPolicy.paid_days ?? 14)
      : Number(expirationPolicy.free_days ?? 1);

    const expiresAt = new Date(
      Date.now() + expiryDays * 24 * 60 * 60 * 1000,
    ).toISOString();

    // ── Step 7: Insert creation ────────────────────────────────────
    const { data: creation, error: insertErr } = await supabase
      .from("creations")
      .insert({
        user_id: user.id,
        template_id: parsed.data.template_id,
        metadata: parsed.data.metadata,
        is_paid: isPaid,
        expires_at: expiresAt,
      })
      .select("id, expires_at")
      .single();

    if (insertErr || !creation) {
      return { error: "Failed to create card", status: 500 };
    }

    // ── Step 8: Decrement creations_left & increment count ─────────
    const updateDict: Record<string, number> = {
      creations_count: (profile.creations_count ?? 0) + 1,
    };

    if (userTier === "free") {
      updateDict.creations_left_free = Math.max(
        (profile.creations_left_free ?? 0) - 1,
        0,
      );
    } else if (
      userTier === "premium" &&
      profile.creations_left_pro !== null
    ) {
      updateDict.creations_left_pro = Math.max(
        (profile.creations_left_pro ?? 0) - 1,
        0,
      );
    }

    await supabase
      .from("profiles")
      .update(updateDict)
      .eq("id", user.id);

    return {
      data: {
        id: creation.id as string,
        expires_at: (creation.expires_at as string) ?? null,
      },
    };
  } catch (e) {
    return {
      error: `Failed to create card: ${e instanceof Error ? e.message : String(e)}`,
      status: 500,
    };
  }
}

// ---------------------------------------------------------------------------
// LIST (authenticated)
// ---------------------------------------------------------------------------

/**
 * GET /creations/my → getMyCreations()
 *
 * Returns current user's non-deleted creations with joined template info.
 * RLS enforces user_id ownership.
 */
export async function getMyCreations(): Promise<
  { data: CreationListItem[] } | { error: string; status: number }
> {
  try {
    const supabase = await createClient();

    const {
      data: { user },
      error: authError,
    } = await supabase.auth.getUser();

    if (authError || !user) {
      return { error: "Unauthorized", status: 401 };
    }

    const { data, error } = await supabase
      .from("creations")
      .select(
        "id, is_paid, expires_at, is_deleted, created_at, templates!inner(slug, name)",
      )
      .eq("user_id", user.id)
      .eq("is_deleted", false)
      .order("created_at", { ascending: false });

    if (error) {
      return { error: error.message, status: 500 };
    }

    const items: CreationListItem[] = (data ?? []).map((c) => {
      const tmpl = (c.templates as unknown as Record<string, string>) ?? {};
      return {
        id: c.id as string,
        template_slug: tmpl.slug ?? "",
        template_name: tmpl.name ?? "כרטיס",
        is_paid: (c.is_paid as boolean) ?? null,
        expires_at: (c.expires_at as string) ?? null,
        is_deleted: (c.is_deleted as boolean) ?? false,
        created_at: c.created_at as string,
      };
    });

    return { data: items };
  } catch (e) {
    return {
      error: `Failed to fetch creations: ${e instanceof Error ? e.message : String(e)}`,
      status: 500,
    };
  }
}

// ---------------------------------------------------------------------------
// GET (public – shared links)
// ---------------------------------------------------------------------------

/**
 * GET /creations/:id → getCreation(creationId)
 *
 * Public endpoint for shared links. Checks is_deleted and expiry.
 * The anon-key RLS policy allows SELECT on non-deleted creations.
 */
export async function getCreation(
  creationId: string,
): Promise<{ data: CreationDetail } | { error: string; status: number }> {
  try {
    const supabase = await createClient();

    const { data: c, error } = await supabase
      .from("creations")
      .select(
        "id, metadata, is_paid, expires_at, is_deleted, created_at, templates!inner(slug, name)",
      )
      .eq("id", creationId)
      .single();

    if (error || !c) {
      return { error: "Creation not found", status: 404 };
    }

    // Deleted check (mirrors Python logic)
    if (c.is_deleted) {
      return { error: "This creation has been deleted", status: 410 };
    }

    // Expiry check (mirrors Python logic)
    if (c.expires_at) {
      try {
        const expStr = String(c.expires_at);
        const expDt = new Date(
          expStr.includes("Z")
            ? expStr
            : expStr.replace("+00:00", "Z"),
        );
        if (expDt < new Date()) {
          return { error: "This creation has expired", status: 410 };
        }
      } catch {
        // Ignore parse errors — same as Python
      }
    }

    const tmpl = (c.templates as unknown as Record<string, string>) ?? {};

    return {
      data: {
        id: c.id as string,
        template_slug: tmpl.slug ?? "",
        template_name: tmpl.name ?? "כרטיס",
        metadata: (c.metadata as Record<string, unknown>) ?? {},
        is_paid: (c.is_paid as boolean) ?? null,
        expires_at: (c.expires_at as string) ?? null,
        created_at: c.created_at as string,
      },
    };
  } catch (e) {
    return {
      error: `Failed to fetch creation: ${e instanceof Error ? e.message : String(e)}`,
      status: 500,
    };
  }
}

// ---------------------------------------------------------------------------
// SOFT DELETE
// ---------------------------------------------------------------------------

/**
 * DELETE /creations/:id → deleteCreation(creationId)
 *
 * Soft-deletes a creation (sets is_deleted = true).
 * RLS enforces that only the owner can modify.
 */
export async function deleteCreation(
  creationId: string,
): Promise<{ success: true } | { error: string; status: number }> {
  try {
    const supabase = await createClient();

    const {
      data: { user },
      error: authError,
    } = await supabase.auth.getUser();

    if (authError || !user) {
      return { error: "Unauthorized", status: 401 };
    }

    // Verify ownership
    const { data, error: findErr } = await supabase
      .from("creations")
      .select("id")
      .eq("id", creationId)
      .eq("user_id", user.id);

    if (findErr || !data?.length) {
      return { error: "Creation not found", status: 404 };
    }

    await supabase
      .from("creations")
      .update({ is_deleted: true })
      .eq("id", creationId);

    return { success: true };
  } catch (e) {
    return {
      error: `Failed to delete creation: ${e instanceof Error ? e.message : String(e)}`,
      status: 500,
    };
  }
}

// ---------------------------------------------------------------------------
// REDEEM COUPON (LoveCoupons template)
// ---------------------------------------------------------------------------

/**
 * Marks a single coupon as redeemed inside the creation's metadata JSONB.
 * No auth required — the viewer (public link recipient) redeems coupons.
 *
 * metadata.coupons[] is an array of { id, title, ..., isRedeemed, redeemedAt }.
 * We find the matching coupon by `couponId`, flip `isRedeemed` to true,
 * and set `redeemedAt` to the current ISO timestamp.
 */
export async function redeemCoupon(
  creationId: string,
  couponId: string,
): Promise<{ success: true } | { error: string; status: number }> {
  try {
    const supabase = await createClient();

    // Fetch the creation's metadata
    const { data, error: fetchErr } = await supabase
      .from("creations")
      .select("metadata")
      .eq("id", creationId)
      .eq("is_deleted", false)
      .single();

    if (fetchErr || !data) {
      return { error: "Creation not found", status: 404 };
    }

    const metadata = (data.metadata ?? {}) as Record<string, unknown>;
    const coupons = Array.isArray(metadata.coupons) ? metadata.coupons : [];

    // Find and mark the coupon
    let found = false;
    const updatedCoupons = coupons.map((c: Record<string, unknown>) => {
      if (c.id === couponId) {
        found = true;
        return { ...c, isRedeemed: true, redeemedAt: new Date().toISOString() };
      }
      return c;
    });

    if (!found) {
      return { error: "Coupon not found", status: 404 };
    }

    // Persist back
    const { error: updateErr } = await supabase
      .from("creations")
      .update({ metadata: { ...metadata, coupons: updatedCoupons } })
      .eq("id", creationId);

    if (updateErr) {
      return {
        error: `Failed to redeem coupon: ${updateErr.message}`,
        status: 500,
      };
    }

    return { success: true };
  } catch (e) {
    return {
      error: `Failed to redeem coupon: ${e instanceof Error ? e.message : String(e)}`,
      status: 500,
    };
  }
}

// ---------------------------------------------------------------------------
// GENERIC CREATION (file upload + insert for ANY template)
// ---------------------------------------------------------------------------

/**
 * submitGenericCreation(formData)
 *
 * Generic server action for creating a card from ANY template.
 * Accepts dynamic templateSlug, metadata JSON, and optional file + bucket.
 *
 * FormData keys:
 *   - templateSlug  (string, required)
 *   - metadata      (JSON string, required)
 *   - file          (File, optional)  — image or asset to upload
 *   - bucketName    (string, optional) — Supabase Storage bucket name
 *
 * Flow:
 *  1. Authenticate user
 *  2. Parse metadata JSON
 *  3. If file + bucketName provided → upload, inject publicUrl into metadata
 *  4. Fetch template_id by slug
 *  5. Quota & premium guard
 *  6. Insert creation row
 *  7. Decrement quota
 *  8. Return { success, creationId }
 */
export async function submitGenericCreation(
  formData: FormData,
): Promise<
  { success: true; creationId: string } | { error: string; status: number }
> {
  try {
    const supabase = await createClient();

    // ── 1. Authenticate ────────────────────────────────────────────
    const {
      data: { user },
      error: authError,
    } = await supabase.auth.getUser();

    if (authError || !user) {
      return { error: "Unauthorized", status: 401 };
    }

    // ── Extract form fields ────────────────────────────────────────
    const templateSlug = formData.get("templateSlug") as string | null;
    const metadataRaw = formData.get("metadata") as string | null;
    const file = formData.get("file") as File | null;
    const bucketName = formData.get("bucketName") as string | null;

    if (!templateSlug?.trim()) {
      return { error: "templateSlug is required", status: 422 };
    }

    if (!metadataRaw) {
      return { error: "metadata is required", status: 422 };
    }

    // ── 2. Parse metadata JSON ─────────────────────────────────────
    let parsedMetadata: Record<string, unknown>;
    try {
      parsedMetadata = JSON.parse(metadataRaw);
    } catch {
      return { error: "metadata must be valid JSON", status: 422 };
    }

    // ── 3. Optional file upload ────────────────────────────────────
    if (file && file.size > 0 && bucketName?.trim()) {
      const fileExt = file.type.split("/")[1] || "jpeg";
      const fileName = `${crypto.randomUUID()}.${fileExt}`;
      const storagePath = `${user.id}/${fileName}`;

      const { error: uploadError } = await supabase.storage
        .from(bucketName)
        .upload(storagePath, file, {
          contentType: file.type,
          upsert: false,
        });

      if (uploadError) {
        console.error("[submitGenericCreation] Upload error", {
          bucketName,
          storagePath,
          fileType: file.type,
          fileSize: file.size,
          error: uploadError,
        });
        return {
          error: `Image upload failed: ${uploadError.message}`,
          status: 500,
        };
      }

      const {
        data: { publicUrl },
      } = supabase.storage.from(bucketName).getPublicUrl(storagePath);

      // Inject the public URL into metadata
      parsedMetadata.background_image = publicUrl;
    }

    // ── 4. Fetch template by slug ──────────────────────────────────
    const { data: template, error: tmplErr } = await supabase
      .from("templates")
      .select("id, is_premium, expiration_policy")
      .eq("slug", templateSlug)
      .eq("is_active", true)
      .single();

    if (tmplErr || !template) {
      return { error: "Template not found", status: 404 };
    }

    // ── 5. Quota & premium guard ───────────────────────────────────
    const { data: profile, error: profErr } = await supabase
      .from("profiles")
      .select(
        "subscription_tier, creations_count, creations_left_free, creations_left_pro",
      )
      .eq("id", user.id)
      .single();

    if (profErr || !profile) {
      return {
        error: "Profile not found. Please complete registration.",
        status: 404,
      };
    }

    const userTier: string = profile.subscription_tier ?? "free";

    if (template.is_premium && userTier === "free") {
      return { error: "TEMPLATE_NOT_ALLOWED", status: 402 };
    }

    if (userTier === "free") {
      const creationsLeft: number = profile.creations_left_free ?? 0;
      if (creationsLeft <= 0) {
        return { error: "QUOTA_EXCEEDED", status: 403 };
      }
    } else if (userTier === "premium") {
      const creationsLeft: number | null = profile.creations_left_pro;
      if (creationsLeft !== null && creationsLeft <= 0) {
        return { error: "QUOTA_EXCEEDED", status: 403 };
      }
    }

    // ── 6. Expiry calculation ──────────────────────────────────────
    const isPaid = userTier === "premium";
    const expirationPolicy =
      (template.expiration_policy as Record<string, unknown>) ?? {};
    const expiryDays = isPaid
      ? Number(expirationPolicy.paid_days ?? 14)
      : Number(expirationPolicy.free_days ?? 1);
    const expiresAt = new Date(
      Date.now() + expiryDays * 24 * 60 * 60 * 1000,
    ).toISOString();

    // ── 7. Insert creation ─────────────────────────────────────────
    const { data: creation, error: insertErr } = await supabase
      .from("creations")
      .insert({
        user_id: user.id,
        template_id: template.id,
        metadata: parsedMetadata,
        is_paid: isPaid,
        expires_at: expiresAt,
      })
      .select("id")
      .single();

    if (insertErr || !creation) {
      return {
        error: `Failed to create card: ${insertErr?.message ?? "Unknown error"}`,
        status: 500,
      };
    }

    // ── 8. Decrement quota & increment count ───────────────────────
    const updateDict: Record<string, number> = {
      creations_count: (profile.creations_count ?? 0) + 1,
    };

    if (userTier === "free") {
      updateDict.creations_left_free = Math.max(
        (profile.creations_left_free ?? 0) - 1,
        0,
      );
    } else if (
      userTier === "premium" &&
      profile.creations_left_pro !== null
    ) {
      updateDict.creations_left_pro = Math.max(
        (profile.creations_left_pro ?? 0) - 1,
        0,
      );
    }

    await supabase.from("profiles").update(updateDict).eq("id", user.id);

    // ── 9. Return success ──────────────────────────────────────────
    return { success: true, creationId: creation.id as string };
  } catch (e) {
    console.error("[submitGenericCreation] Unexpected error", e);
    return {
      error: `Failed to create card: ${e instanceof Error ? e.message : String(e)}`,
      status: 500,
    };
  }
}
